aten/src/ATen/native/cpu/SpmmReduceKernel.cpp

```diff
@@ -21,6 +21,135 @@ namespace at { namespace native {
 
 namespace {
 
+template <typename scalar_t, typename index_t, ReductionType reduce>
+inline void _update(scalar_t* out_ptr, int64_t row_start, int64_t row_end, const TensorAccessor<index_t, 1>& col_data, const TensorAccessor<scalar_t, 1>& val_data, scalar_t* other_data, int64_t K) {
+  using Vec = vec::Vectorized<scalar_t>;
+  using fVec = vec::Vectorized<float>;
+  constexpr int64_t kVecSize = Vec::size();
+  constexpr int64_t kVLEN = kVecSize * 4;
+  constexpr int64_t CHUNK_SIZE = 16;
+
+  if constexpr (!is_reduced_floating_point_v<scalar_t>) {
+    for (int64_t e0 = row_start; e0 < row_end; e0 += CHUNK_SIZE) {
+      int64_t e1 = std::min(e0 + CHUNK_SIZE, row_end);
+
+      int64_t k = 0, c;
+      for (; k < K - (K % kVLEN); k += kVLEN) {
+        Vec out_vec0 = Vec::loadu(out_ptr + k);
+        Vec out_vec1 = Vec::loadu(out_ptr + k + kVecSize);
+        Vec out_vec2 = Vec::loadu(out_ptr + k + kVecSize * 2);
+        Vec out_vec3 = Vec::loadu(out_ptr + k + kVecSize * 3);
+        for (const auto e : c10::irange(e0, e1)) {
+          c = col_data[e];
+          scalar_t val = val_data[e];
+          scalar_t* other_ptr = other_data + c * K + k;
+
+          out_vec0 = update<Vec, reduce>(out_vec0, Vec::loadu(other_ptr) * Vec(val));
+          out_vec1 = update<Vec, reduce>(out_vec1, Vec::loadu(other_ptr + kVecSize) * Vec(val));
+          out_vec2 = update<Vec, reduce>(out_vec2, Vec::loadu(other_ptr + kVecSize * 2) * Vec(val));
+          out_vec3 = update<Vec, reduce>(out_vec3, Vec::loadu(other_ptr + kVecSize * 3) * Vec(val));
+        }
+        out_vec0.store(out_ptr + k);
+        out_vec1.store(out_ptr + k + kVecSize);
+        out_vec2.store(out_ptr + k + kVecSize * 2);
+        out_vec3.store(out_ptr + k + kVecSize * 3);
+      }
+      for (; k < K - (K % kVecSize); k += kVecSize) {
+        Vec out_vec = Vec::loadu(out_ptr + k);
+        for (const auto e : c10::irange(e0, e1)) {
+          c = col_data[e];
+          scalar_t val = val_data[e];
+          scalar_t* other_ptr = other_data + c * K;
+          out_vec = update<Vec, reduce>(out_vec, Vec::loadu(other_ptr + k) * Vec(val));
+        }
+        out_vec.store(out_ptr + k);
+      }
+      for (; k < K; k++) {
+        scalar_t out_val = out_ptr[k];
+        for (const auto e : c10::irange(e0, e1)) {
+          c = col_data[e];
+          scalar_t val = val_data[e];
+          scalar_t* other_ptr = other_data + c * K;
+          out_val = update<scalar_t, reduce>(out_val, other_ptr[k] * val);
+        }
+        out_ptr[k] = out_val;
+      }
+    }
+  } else {
+    for (int64_t e0 = row_start; e0 < row_end; e0 += CHUNK_SIZE) {
+      int64_t e1 = std::min(e0 + CHUNK_SIZE, row_end);
+
+      int64_t k = 0, c;
+      for (; k < K - (K % kVLEN); k += kVLEN) {
+        Vec out_vec0 = Vec::loadu(out_ptr + k);
+        fVec out_vec0_0, out_vec0_1;
+        std::tie(out_vec0_0, out_vec0_1) = convert_to_float<scalar_t>(out_vec0);
```

Comment:
consider to use c++17 unpacking: `auto [out_vec0_0, out_vec0_1]`

Reply:
After refactor code using VecType<scalar_t>, the unpacking is gone.